src/Features/Core/Portable/EmbeddedLanguages/RegularExpressions/RegexParser.cs

```diff
@@ -251,26 +252,38 @@ private RegexExpressionNode ParseAlternatingSequences(bool consumeCloseParen)
         }
 
         /// <summary>
-        /// Parses out code of the form: ...|...|...
-        /// This is the type of code you have at the top level of a regex, or inside any grouping
-        /// contruct.  Note that sequences can be empty in .NET regex.  i.e. the following is legal:
-        /// 
+        /// Parses out code of the form: ...|...|... This is the type of code you have at the top level of a regex, or
+        /// inside any grouping construct.  Note that sequences can be empty in .NET regex.  i.e. the following is
+        /// legal:
+        ///
         ///     ...||...
-        /// 
+        ///
         /// An empty sequence just means "match at every position in the test string".
         /// </summary>
-        private RegexExpressionNode ParseAlternatingSequencesWorker(bool consumeCloseParen)
+        private RegexAlternationNode ParseAlternatingSequencesWorker(
+            bool consumeCloseParen, bool isConditional)
         {
-            RegexExpressionNode current = ParseSequence(consumeCloseParen);
+            using var _ = ArrayBuilder<RegexNodeOrToken>.GetInstance(out var builder);
+            builder.Add(ParseSequence(consumeCloseParen));
 
             while (_currentToken.Kind == RegexKind.BarToken)
             {
                 // Trivia allowed between the | and the next token.
-                current = new RegexAlternationNode(
-                    current, ConsumeCurrentToken(allowTrivia: true), ParseSequence(consumeCloseParen));
+                var barToken = ConsumeCurrentToken(allowTrivia: true);
+                if (isConditional && builder.Count >= 3)
+                {
+                    // a conditional alternative expression only allows two cases (the true and false branches). We
+                    // already have seen both once we have 3 items (`t | f`).  Error on any further cases we see.
+                    barToken = barToken.AddDiagnosticIfNone(new EmbeddedDiagnostic(
+                        FeaturesResources.Too_many_bars_in_conditional_grouping,
+                        barToken.GetSpan()));
+                }
```

Comment:
For `"(?(0)a|b|c|d|e|f)"`, were we previously reporting a single diagnostic but now multiple diagnostics?

Reply:
i'm ok with that.  i'll write a test to this effect.

Reply:
tests added!